src/System.Private.ServiceModel/tests/Stress/SharedPoolsOfWCFObjects/SharedPoolsOfWCFObjects.cs

```diff
@@ -0,0 +1,612 @@
+        private static CallStats s_CallChannelStats = new CallStats(samples: 1000000, errorSamples: 10000);
+        private static CallStats s_CloseChannelStats = new CallStats(samples: 1000, errorSamples: 1000);
+        private static CallStats s_CloseFactoryStats = new CallStats(samples: 1000, errorSamples: 1000);
+        private static PoolOfThings<FactoryAndPoolOfItsObjects<
+                ChannelFactory<WcfService1.IService1>,
+                OpenOnceChannelWrapper<WcfService1.IService1>>
+            > s_recyclablePooledFactoriesAndChannels;
+
+        static RecyclablePooledFactoriesAndChannels_OpenOnce()
+        {
+            s_recyclablePooledFactoriesAndChannels = StaticDisposablesHelper.AddDisposable(
+                new PoolOfThings<FactoryAndPoolOfItsObjects<ChannelFactory<WcfService1.IService1>, OpenOnceChannelWrapper<WcfService1.IService1>>>(
+                    maxSize: 3, // # of pooled FactoryAndPoolOfItsObjects
+                    createInstance: () => new FactoryAndPoolOfItsObjects<ChannelFactory<WcfService1.IService1>, OpenOnceChannelWrapper<WcfService1.IService1>>(
+                        createFactoryInstance: () =>
+                            TestHelpers.CreateChannelFactory<WcfService1.IService1>(TestHelpers.CreateEndPointAddress(), TestHelpers.CreateBinding()),
+                        destroyFactoryInstance: (chf) =>
+                            s_CloseFactoryStats.CallActionAndRecordStats(() => TestHelpers.CloseFactory(chf)),
+                        maxPooledObjects: 3, // # of pooled channels within each pooled FactoryAndPoolOfItsObjects
+                        createObject: (chf) =>
+                            s_CreateChannelStats.CallFuncAndRecordStats(func: () => new OpenOnceChannelWrapper<WcfService1.IService1>(TestHelpers.CreateChannel(chf))),
+                        destroyObject: (chWr) =>
+                            s_CloseChannelStats.CallActionAndRecordStats(action: () => TestHelpers.CloseChannel(chWr.Channel))
+                        ),
+                    destroyInstance: (_fapoic) => _fapoic.Destroy()));
+        }
+
+        public static void RunAllScenariosWithWeights(int useWeight, int recycleChannelsWeight, int recycleFactoriesWeight)
+        {
+            int seed = new Random(Interlocked.Increment(ref s_iteration)).Next(useWeight + recycleChannelsWeight + recycleFactoriesWeight);
+            if (seed < useWeight)
+            {
+                UsePooledChannels();
+            }
+            else if (seed < useWeight + recycleChannelsWeight)
+            {
+                RecyclePooledChannels();
+            }
+            else if (seed < useWeight + recycleChannelsWeight + recycleFactoriesWeight)
+            {
+                RecyclePooledFactories();
+            }
+
+            // move this outside of the test - an external code should query the test for its call stats
+            if (Interlocked.Increment(ref s_iteration) % 1000000 == 0)
+            {
```

Comment:
For future consideration -- and I'll stop commenting on them.  Magic literals like 1000000 should be some visible const or static field.  Also, the comment counts as a secret TODO, so it should be turned into an issue if you really intend to do it.